includes/Modules/Analytics_4/Report/Response.php

```diff
@@ -0,0 +1,138 @@
+			return $response;
+		}
+
+		// Get date ranges and return early if there are no date ranges for this report.
+		$date_ranges = $this->parse_dateranges( $data );
+		if ( empty( $date_ranges ) ) {
+			return $response;
+		}
+
+		// Get all available dates in the report.
+		$existing_rows = array();
+		foreach ( $response->getRows() as $row ) {
+			$dimension_values = $row->getDimensionValues();
+
+			$range = 'date_range_0';
+			if ( count( $dimension_values ) > 1 ) {
+				$range = $dimension_values[1]->getValue();
+			}
+
+			$range = str_replace( 'date_range_', '', $range );
+			$date  = $dimension_values[0]->getValue();
+			$key   = self::get_response_row_key( $date, is_numeric( $range ) ? $range : false );
+
+			$existing_rows[ $key ] = $row;
+		}
+
+		$metric_headers  = $response->getMetricHeaders();
+		$ranges_count    = count( $date_ranges );
+		$multiple_ranges = $ranges_count > 1;
+		$rows            = array();
+
+		foreach ( $date_ranges as $date_range ) {
+			$start = strtotime( $date_range->getStartDate() );
+			$end   = strtotime( $date_range->getEndDate() );
+
+			// Skip this date range if either start date or end date is corrupted.
+			if ( ! $start || ! $end ) {
+				continue;
+			}
+
+			// Loop through all days in the date range and check if there is a metric value
+			// for it. If the metric value is missing, we will need to add one with a zero value.
+			$now = $start;
+			do {
+				// Format the current time to a date string and add a day in seconds to the current date
+				// to shift to the next date.
+				$current_date = gmdate( 'Ymd', $now );
+				$now         += DAY_IN_SECONDS;
+
+				// Add rows for the current date for each date range.
+				for ( $i = 0; $i < $ranges_count; $i++ ) {
+					// Copy the existing row if it is available, otherwise create a new zero-value row.
+					$current_date_key          = self::get_response_row_key( $current_date, $i );
+					$rows[ $current_date_key ] = isset( $existing_rows[ $current_date_key ] )
+						? $existing_rows[ $current_date_key ]
+						: $this->create_report_row(
+							$metric_headers,
+							$current_date,
+							$multiple_ranges ? $i : false
+						);
+				}
+			} while ( $now <= $end );
+		}
+
+		// If we have the same number of rows as in the response at the moment, then
+		// we can return the response without setting the new rows back into the response.
+		$new_rows_count = count( $rows );
+		if ( $new_rows_count <= $response->getRowCount() ) {
+			return $response;
+		}
+
+		// Set updated rows back to the response object.
+		$response->setRows( array_values( $rows ) );
+		$response->setRowCount( $new_rows_count );
+
+		return $response;
+	}
+
+	/**
+	 * Gets the response row key composed from the date and the date range index values.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $date             The date of the row to return key for.
+	 * @param int    $date_range_index The date range index.
```

Comment:
Considering how it's invoked, `$date_range_index` could in theory be `false`, although in practice this scenario shouldn't actually occur.

```suggestion
	 * @param int|bool    $date_range_index The date range index, or `false` if no index is available.
```